# Pandas Methods()

In [2]:
# initializing pandas
import pandas as pd

In [3]:
# pandas ingestion
df_hr = pd.read_csv("data/HR_comma_sep.csv")
df_hr

satisfaction_level  last_evaluation  number_project  \
0                    0.38             0.53               2   
1                    0.80             0.86               5   
2                    0.11             0.88               7   
3                    0.72             0.87               5   
4                    0.37             0.52               2   
...                   ...              ...             ...   
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
0                       157                   3              0     1   
1                       262                   6              0     1   
2                       272                   4              0     1   
3                       223                   5              0     1   
4                       159                   3              0     1   
...                     ...                 ...            ...   ...   
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department  salary  
0                          0      sales     low  
1                          0      sales  medium  
2                          0      sales  medium  
3                          0      sales     low  
4                          0      sales     low  
...                      ...        ...     ...  
14994                      0    support     low  
14995                      0    support     low  
14996                      0    support     low  
14997                      0    support     low  
14998                      0    support     low  

[14999 rows x 10 columns]

# .head()

In [4]:
# use .head() to review first five data 
df_hr.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [5]:
# use .head(10) to review first 10 data 
df_hr.head(10)

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   
5                0.41             0.50               2                   153   
6                0.10             0.77               6                   247   
7                0.92             0.85               5                   259   
8                0.89             1.00               5                   224   
9                0.42             0.53               2                   142   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   
5                   3              0     1                      0      sales   
6                   4              0     1                      0      sales   
7                   5              0     1                      0      sales   
8                   5              0     1                      0      sales   
9                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low  
5     low  
6     low  
7     low  
8     low  
9     low

# .tail()

In [6]:
# use .tail() to review last five data
df_hr.tail()

satisfaction_level  last_evaluation  number_project  \
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department salary  
14994                      0    support    low  
14995                      0    support    low  
14996                      0    support    low  
14997                      0    support    low  
14998                      0    support    low

In [7]:
# use .tail(10) to review last 10 data
df_hr.tail(10)

satisfaction_level  last_evaluation  number_project  \
14989                0.43             0.57               2   
14990                0.89             0.88               5   
14991                0.09             0.81               6   
14992                0.40             0.48               2   
14993                0.76             0.83               6   
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
14989                   159                   3              1     1   
14990                   228                   5              1     1   
14991                   257                   4              0     1   
14992                   155                   3              0     1   
14993                   293                   6              0     1   
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department salary  
14989                      0  technical    low  
14990                      0    support    low  
14991                      0    support    low  
14992                      0    support    low  
14993                      0    support    low  
14994                      0    support    low  
14995                      0    support    low  
14996                      0    support    low  
14997                      0    support    low  
14998                      0    support    low

# .shape

In [8]:
# use .shape to view the shape of data
df_hr.shape

(14999, 10)

# .index

In [9]:
# use .index to view the index of data
df_hr.index

RangeIndex(start=0, stop=14999, step=1)

# .columns

In [10]:
# use .columns to view the data of column
df_hr.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'Department', 'salary'],
      dtype='object')

# .dtypes

In [11]:
# use .dtypes to classify data type of data
df_hr.dtypes

satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
left                       int64
promotion_last_5years      int64
Department                object
salary                    object
dtype: object

In [12]:
df_new = df_hr

In [13]:
df_new

satisfaction_level  last_evaluation  number_project  \
0                    0.38             0.53               2   
1                    0.80             0.86               5   
2                    0.11             0.88               7   
3                    0.72             0.87               5   
4                    0.37             0.52               2   
...                   ...              ...             ...   
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
0                       157                   3              0     1   
1                       262                   6              0     1   
2                       272                   4              0     1   
3                       223                   5              0     1   
4                       159                   3              0     1   
...                     ...                 ...            ...   ...   
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department  salary  
0                          0      sales     low  
1                          0      sales  medium  
2                          0      sales  medium  
3                          0      sales     low  
4                          0      sales     low  
...                      ...        ...     ...  
14994                      0    support     low  
14995                      0    support     low  
14996                      0    support     low  
14997                      0    support     low  
14998                      0    support     low  

[14999 rows x 10 columns]

# .rename() 

In [14]:
# use .rename() to rename the column name
df_new.rename(columns = {'last_evaluation': 'kxa'}, inplace=True)

In [15]:
df_new

satisfaction_level   kxa  number_project  average_montly_hours  \
0                    0.38  0.53               2                   157   
1                    0.80  0.86               5                   262   
2                    0.11  0.88               7                   272   
3                    0.72  0.87               5                   223   
4                    0.37  0.52               2                   159   
...                   ...   ...             ...                   ...   
14994                0.40  0.57               2                   151   
14995                0.37  0.48               2                   160   
14996                0.37  0.53               2                   143   
14997                0.11  0.96               6                   280   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident  left  promotion_last_5years  \
0                       3              0     1                      0   
1                       6              0     1                      0   
2                       4              0     1                      0   
3                       5              0     1                      0   
4                       3              0     1                      0   
...                   ...            ...   ...                    ...   
14994                   3              0     1                      0   
14995                   3              0     1                      0   
14996                   3              0     1                      0   
14997                   4              0     1                      0   
14998                   3              0     1                      0   

      Department  salary  
0          sales     low  
1          sales  medium  
2          sales  medium  
3          sales     low  
4          sales     low  
...          ...     ...  
14994    support     low  
14995    support     low  
14996    support     low  
14997    support     low  
14998    support     low  

[14999 rows x 10 columns]

# .values

In [16]:
# .values to view all value of dataset
df_hr.values

array([[0.38, 0.53, 2, ..., 0, 'sales', 'low'],
       [0.8, 0.86, 5, ..., 0, 'sales', 'medium'],
       [0.11, 0.88, 7, ..., 0, 'sales', 'medium'],
       ...,
       [0.37, 0.53, 2, ..., 0, 'support', 'low'],
       [0.11, 0.96, 6, ..., 0, 'support', 'low'],
       [0.37, 0.52, 2, ..., 0, 'support', 'low']], dtype=object)

# .info()

In [17]:
# use .info() to view all information of dataset
df_hr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   kxa                    14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   Department             14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


* DataFrame occuopies 1.1+ MB memory
* Columns are of types: float64, int64, object
* Total columns = 14999
* No any columns contain null values

# .describe()

In [18]:
# use .describe() to view numerical operation information of dataset
df_hr.describe()

satisfaction_level           kxa  number_project  average_montly_hours  \
count        14999.000000  14999.000000    14999.000000          14999.000000   
mean             0.612834      0.716102        3.803054            201.050337   
std              0.248631      0.171169        1.232592             49.943099   
min              0.090000      0.360000        2.000000             96.000000   
25%              0.440000      0.560000        3.000000            156.000000   
50%              0.640000      0.720000        4.000000            200.000000   
75%              0.820000      0.870000        5.000000            245.000000   
max              1.000000      1.000000        7.000000            310.000000   

       time_spend_company  Work_accident          left  promotion_last_5years  
count        14999.000000   14999.000000  14999.000000           14999.000000  
mean             3.498233       0.144610      0.238083               0.021268  
std              1.460136       0.351719      0.425924               0.144281  
min              2.000000       0.000000      0.000000               0.000000  
25%              3.000000       0.000000      0.000000               0.000000  
50%              3.000000       0.000000      0.000000               0.000000  
75%              4.000000       0.000000      0.000000               0.000000  
max             10.000000       1.000000      1.000000               1.000000

* We get the summary of each dataframe column easily without the need for explicit computation
* By default it includes only the numerical columns.
* let's get description of object type only below cells

In [19]:
# use .describe(include='object') to view object operation information of dataset
df_hr.describe(include='object')

Department salary
count       14999  14999
unique         10      3
top         sales    low
freq         4140   7316

* For object type, we get count, uinque, top item with frequency

#  Data Selection
* The section covers differnt techniques to access columns,slice dataframe, get unique column values.
    * select column(using[], and dot(.) operator)
    * iloc
    * loc
    * unique() and nunique()
    * value.counts()

In [20]:
# selecting single column
df_hr["satisfaction_level"]

0        0.38
1        0.80
2        0.11
3        0.72
4        0.37
         ... 
14994    0.40
14995    0.37
14996    0.37
14997    0.11
14998    0.37
Name: satisfaction_level, Length: 14999, dtype: float64

In [21]:
# selecting multiple columns
column_to_select = ['satisfaction_level','left']
df_hr[column_to_select]

satisfaction_level  left
0                    0.38     1
1                    0.80     1
2                    0.11     1
3                    0.72     1
4                    0.37     1
...                   ...   ...
14994                0.40     1
14995                0.37     1
14996                0.37     1
14997                0.11     1
14998                0.37     1

[14999 rows x 2 columns]

# .iloc
* ways ti slice pandas dataframe via integer value.
* it allows you ti select data from a DataFrame using integer indices for both rows and columns


In [22]:
# use .iloc to slicing
df_hr.iloc[0,0]

0.38

In [23]:
df_hr.iloc[:3,-3:]

promotion_last_5years Department  salary
0                      0      sales     low
1                      0      sales  medium
2                      0      sales  medium

# .loc
* loc is used for label-based indexing
* allows us to slic DataFrame using row label and column label instead os integer value as in loc

In [24]:
df_hr.loc[0, 'satisfaction_level']

0.38

In [25]:
row_label = [0,1]
column_label = ['promotion_last_5years', 'Department', 'salary']
df_hr.loc[row_label, column_label]

promotion_last_5years Department  salary
0                      0      sales     low
1                      0      sales  medium

# unique() and nunique()
* unique() use to view the unique item of any column
* nunique() use to view the number of unique item of any column

In [26]:
# use to view unique value of "left" column
df_hr['left'].unique()

array([1, 0], dtype=int64)

In [27]:
# use to view number of unique value of "left" column
df_hr['left'].nunique()

2

# value_counts()

In [28]:
# value_counts() use to count the number of unique value
df_hr['left'].value_counts()

left
0    11428
1     3571
Name: count, dtype: int64

# Aggregation
* Methods
  * mean()
  * median()
  * std()
  * sum()
  * max()
  * min()

In [29]:
df_hr["satisfaction_level"].mean()

0.6128335222348156

In [30]:
df_hr["satisfaction_level"].median()

0.64

In [31]:
df_hr["satisfaction_level"].std()

0.24863065106114257

In [32]:
df_hr["satisfaction_level"].sum()

9191.89

In [33]:
df_hr["satisfaction_level"].min()


0.09

# Missing Data
* Methods:
  * isnull() or isna()
  * notnull() or notna()
  * dropna()
  * fillna()

# isnull() isna()
* Return True if DataFrame values are null
* DataFrame methods to get nan values from columns.
* it will considered null values it there are nan which is python default marker for null values.
* you can use df.isnull() methods which is simply an alias of df,isna()

In [34]:
df_hr.isnull()

satisfaction_level    kxa  number_project  average_montly_hours  \
0                   False  False           False                 False   
1                   False  False           False                 False   
2                   False  False           False                 False   
3                   False  False           False                 False   
4                   False  False           False                 False   
...                   ...    ...             ...                   ...   
14994               False  False           False                 False   
14995               False  False           False                 False   
14996               False  False           False                 False   
14997               False  False           False                 False   
14998               False  False           False                 False   

       time_spend_company  Work_accident   left  promotion_last_5years  \
0                   False          False  False                  False   
1                   False          False  False                  False   
2                   False          False  False                  False   
3                   False          False  False                  False   
4                   False          False  False                  False   
...                   ...            ...    ...                    ...   
14994               False          False  False                  False   
14995               False          False  False                  False   
14996               False          False  False                  False   
14997               False          False  False                  False   
14998               False          False  False                  False   

       Department  salary  
0           False   False  
1           False   False  
2           False   False  
3           False   False  
4           False   False  
...           ...     ...  
14994       False   False  
14995       False   False  
14996       False   False  
14997       False   False  
14998       False   False  

[14999 rows x 10 columns]

In [35]:
df_hr.isnull().sum()

satisfaction_level       0
kxa                      0
number_project           0
average_montly_hours     0
time_spend_company       0
Work_accident            0
left                     0
promotion_last_5years    0
Department               0
salary                   0
dtype: int64

# This mean there is not any missing value in our dataset

In [36]:
df_hr.notnull().sum(axis=0)

satisfaction_level       14999
kxa                      14999
number_project           14999
average_montly_hours     14999
time_spend_company       14999
Work_accident            14999
left                     14999
promotion_last_5years    14999
Department               14999
salary                   14999
dtype: int64

In [37]:
df_hr.dropna()

satisfaction_level   kxa  number_project  average_montly_hours  \
0                    0.38  0.53               2                   157   
1                    0.80  0.86               5                   262   
2                    0.11  0.88               7                   272   
3                    0.72  0.87               5                   223   
4                    0.37  0.52               2                   159   
...                   ...   ...             ...                   ...   
14994                0.40  0.57               2                   151   
14995                0.37  0.48               2                   160   
14996                0.37  0.53               2                   143   
14997                0.11  0.96               6                   280   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident  left  promotion_last_5years  \
0                       3              0     1                      0   
1                       6              0     1                      0   
2                       4              0     1                      0   
3                       5              0     1                      0   
4                       3              0     1                      0   
...                   ...            ...   ...                    ...   
14994                   3              0     1                      0   
14995                   3              0     1                      0   
14996                   3              0     1                      0   
14997                   4              0     1                      0   
14998                   3              0     1                      0   

      Department  salary  
0          sales     low  
1          sales  medium  
2          sales  medium  
3          sales     low  
4          sales     low  
...          ...     ...  
14994    support     low  
14995    support     low  
14996    support     low  
14997    support     low  
14998    support     low  

[14999 rows x 10 columns]

# fillna()
* filling missing values

In [38]:
df_hr.fillna(5)

satisfaction_level   kxa  number_project  average_montly_hours  \
0                    0.38  0.53               2                   157   
1                    0.80  0.86               5                   262   
2                    0.11  0.88               7                   272   
3                    0.72  0.87               5                   223   
4                    0.37  0.52               2                   159   
...                   ...   ...             ...                   ...   
14994                0.40  0.57               2                   151   
14995                0.37  0.48               2                   160   
14996                0.37  0.53               2                   143   
14997                0.11  0.96               6                   280   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident  left  promotion_last_5years  \
0                       3              0     1                      0   
1                       6              0     1                      0   
2                       4              0     1                      0   
3                       5              0     1                      0   
4                       3              0     1                      0   
...                   ...            ...   ...                    ...   
14994                   3              0     1                      0   
14995                   3              0     1                      0   
14996                   3              0     1                      0   
14997                   4              0     1                      0   
14998                   3              0     1                      0   

      Department  salary  
0          sales     low  
1          sales  medium  
2          sales  medium  
3          sales     low  
4          sales     low  
...          ...     ...  
14994    support     low  
14995    support     low  
14996    support     low  
14997    support     low  
14998    support     low  

[14999 rows x 10 columns]

# Data Manipulation
* Methods:
    * replace()
      * replace values from DataFrame
        * df.replace({'old_Value':'new_value'})
    * astype()
      * change DataFrame column types
        * df.astype({'col1': name_type})
    * drop()
    * sort_values()
    * groupby()
    * apply()
    

In [39]:
df_hr['left'].replace({1:"yes",0:"no"})

0        yes
1        yes
2        yes
3        yes
4        yes
        ... 
14994    yes
14995    yes
14996    yes
14997    yes
14998    yes
Name: left, Length: 14999, dtype: object

In [40]:
df_hr.astype({'left': bool})

satisfaction_level   kxa  number_project  average_montly_hours  \
0                    0.38  0.53               2                   157   
1                    0.80  0.86               5                   262   
2                    0.11  0.88               7                   272   
3                    0.72  0.87               5                   223   
4                    0.37  0.52               2                   159   
...                   ...   ...             ...                   ...   
14994                0.40  0.57               2                   151   
14995                0.37  0.48               2                   160   
14996                0.37  0.53               2                   143   
14997                0.11  0.96               6                   280   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident  left  promotion_last_5years  \
0                       3              0  True                      0   
1                       6              0  True                      0   
2                       4              0  True                      0   
3                       5              0  True                      0   
4                       3              0  True                      0   
...                   ...            ...   ...                    ...   
14994                   3              0  True                      0   
14995                   3              0  True                      0   
14996                   3              0  True                      0   
14997                   4              0  True                      0   
14998                   3              0  True                      0   

      Department  salary  
0          sales     low  
1          sales  medium  
2          sales  medium  
3          sales     low  
4          sales     low  
...          ...     ...  
14994    support     low  
14995    support     low  
14996    support     low  
14997    support     low  
14998    support     low  

[14999 rows x 10 columns]

In [41]:
df_hr.drop('salary', axis =1)

satisfaction_level   kxa  number_project  average_montly_hours  \
0                    0.38  0.53               2                   157   
1                    0.80  0.86               5                   262   
2                    0.11  0.88               7                   272   
3                    0.72  0.87               5                   223   
4                    0.37  0.52               2                   159   
...                   ...   ...             ...                   ...   
14994                0.40  0.57               2                   151   
14995                0.37  0.48               2                   160   
14996                0.37  0.53               2                   143   
14997                0.11  0.96               6                   280   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident  left  promotion_last_5years  \
0                       3              0     1                      0   
1                       6              0     1                      0   
2                       4              0     1                      0   
3                       5              0     1                      0   
4                       3              0     1                      0   
...                   ...            ...   ...                    ...   
14994                   3              0     1                      0   
14995                   3              0     1                      0   
14996                   3              0     1                      0   
14997                   4              0     1                      0   
14998                   3              0     1                      0   

      Department  
0          sales  
1          sales  
2          sales  
3          sales  
4          sales  
...          ...  
14994    support  
14995    support  
14996    support  
14997    support  
14998    support  

[14999 rows x 9 columns]

In [42]:
df_hr.sort_values(by="Work_accident", ascending = False)

satisfaction_level   kxa  number_project  average_montly_hours  \
8078                 0.77  0.91               5                   268   
3570                 0.50  0.60               3                   270   
1784                 0.43  0.51               2                   149   
6589                 0.64  0.64               3                   234   
11800                0.67  0.91               2                   245   
...                   ...   ...             ...                   ...   
5310                 0.63  0.87               5                   248   
5311                 0.86  0.75               5                   157   
5312                 0.80  0.79               5                   240   
5313                 0.55  0.58               5                   262   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident  left  promotion_last_5years  \
8078                    3              1     0                      0   
3570                    3              1     0                      0   
1784                    3              1     1                      0   
6589                    3              1     0                      0   
11800                   2              1     0                      1   
...                   ...            ...   ...                    ...   
5310                    2              0     0                      0   
5311                    4              0     0                      0   
5312                    2              0     0                      0   
5313                    3              0     0                      0   
14998                   3              0     1                      0   

       Department  salary  
8078           IT  medium  
3570   management  medium  
1784   accounting  medium  
6589        sales    high  
11800       sales  medium  
...           ...     ...  
5310    technical     low  
5311      support     low  
5312      support     low  
5313      support  medium  
14998     support     low  

[14999 rows x 10 columns]

In [43]:
df_hr.groupby(['Department', 'promotion_last_5years']).sum()

satisfaction_level      kxa  \
Department  promotion_last_5years                                
IT          0                                  757.10   878.10   
            1                                    1.36     1.45   
RandD       0                                  470.48   540.27   
            1                                   17.32    20.17   
accounting  0                                  437.97   539.70   
            1                                    8.54    10.79   
hr          0                                  432.99   514.35   
            1                                    9.53     9.49   
management  0                                  349.22   407.73   
            1                                   42.23    48.39   
marketing   0                                  502.66   584.87   
            1                                   28.10    29.36   
product_mng 0                                  558.91   644.71   
sales       0                                 2472.96  2865.69   
            1                                   70.85    72.54   
support     0                                 1365.36  1598.63   
            1                                   12.83    13.18   
technical   0                                 1634.97  1941.54   
            1                                   18.51    19.85   

                                   number_project  average_montly_hours  \
Department  promotion_last_5years                                         
IT          0                                4677                247715   
            1                                   6                   404   
RandD       0                                2923                152986   
            1                                 110                  5044   
accounting  0                                2874                151521   
            1                                  60                  2771   
hr          0                                2650                143589   
            1                                  51                  3239   
management  0                                2181                113541   
            1                                 251                 13246   
marketing   0                                3016                162663   
            1                                 148                  8410   
product_mng 0                                3434                180369   
sales       0                               15255                811061   
            1                                 379                 20712   
support     0                                8403                443451   
            1                                  76                  4039   
technical   0                               10432                544906   
            1                                 116                  5887   

                                   time_spend_company  Work_accident  left  \
Department  promotion_last_5years                                            
IT          0                                    4247            164   270   
            1                                       9              0     3   
RandD       0                                    2566            128   121   
            1                                      84              6     0   
accounting  0                                    2631             95   204   
            1                                      71              1     0   
hr          0                                    2400             84   215   
            1                                      80              5     0   
management  0                                    2314             86    88   
            1                                     397             17     3   
marketing   0                                    2890            129   203   
            1                                     173     

## Data Join
- `Types are:`
    1. Inner Join
        - pd.merge(df1, df2, on='col_name', how='inner')
    2. Left Outer Join
        - pd.merge(df1, df2, on='col_name', how='left')
    3. Right Outer Join
        - pd.merge(df1, df2, on='col_name', how='right')
    4. Full Outer Join
        - pd.merge(df1, df2, on='col_name', how='outer')

In [44]:
# reset index to get employee id

df_hr_ = df_hr.reset_index()

# rename col index to emp_id
df_hr_.columns.values[0] = "emp_id"

df_hr_.head()



emp_id  satisfaction_level   kxa  number_project  average_montly_hours  \
0       0                0.38  0.53               2                   157   
1       1                0.80  0.86               5                   262   
2       2                0.11  0.88               7                   272   
3       3                0.72  0.87               5                   223   
4       4                0.37  0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [45]:
df_depart = df_hr_[['emp_id','Department']]
df_hr_.drop('Department', axis=1, inplace=True)
df_hr_.head()

emp_id  satisfaction_level   kxa  number_project  average_montly_hours  \
0       0                0.38  0.53               2                   157   
1       1                0.80  0.86               5                   262   
2       2                0.11  0.88               7                   272   
3       3                0.72  0.87               5                   223   
4       4                0.37  0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years  salary  
0                   3              0     1                      0     low  
1                   6              0     1                      0  medium  
2                   4              0     1                      0  medium  
3                   5              0     1                      0     low  
4                   3              0     1                      0     low

# 1. Inner join

In [46]:
# inner join based on employee id to get Department information

df_inner = pd.merge(df_hr_, df_depart, on='emp_id', how='inner')

df_inner.head()

emp_id  satisfaction_level   kxa  number_project  average_montly_hours  \
0       0                0.38  0.53               2                   157   
1       1                0.80  0.86               5                   262   
2       2                0.11  0.88               7                   272   
3       3                0.72  0.87               5                   223   
4       4                0.37  0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years  salary  \
0                   3              0     1                      0     low   
1                   6              0     1                      0  medium   
2                   4              0     1                      0  medium   
3                   5              0     1                      0     low   
4                   3              0     1                      0     low   

  Department  
0      sales  
1      sales  
2      sales  
3      sales  
4      sales